server/neptune/gateway/event/preworkflow/hook_runner.go

```diff
@@ -0,0 +1,54 @@
+package preworkflow
+
+import (
+	"github.com/pkg/errors"
+	"github.com/runatlantis/atlantis/server/core/config/valid"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"github.com/runatlantis/atlantis/server/neptune/gateway/event/source"
+)
+
+type HooksRunner interface {
+	Run(repo models.Repo, sha string) (string, error)
+}
+
+// PreWorkflowHooksRunner is the first step when processing a workflow hook commands.
+type PreWorkflowHooksRunner struct {
+	WorkingDir   source.TmpWorkingDir
+	GlobalCfg    valid.GlobalCfg
+	HookExecutor HookExecutor
+}
+
+func (r *PreWorkflowHooksRunner) Run(baseRepo models.Repo, sha string) (string, error) {
+	preWorkflowHooks := make([]*valid.PreWorkflowHook, 0)
+	for _, repo := range r.GlobalCfg.Repos {
+		if repo.IDMatches(baseRepo.ID()) && len(repo.PreWorkflowHooks) > 0 {
+			preWorkflowHooks = append(preWorkflowHooks, repo.PreWorkflowHooks...)
+		}
+	}
+
+	// short circuit any other calls if there are no pre-hooks configured
+	if len(preWorkflowHooks) == 0 {
```

Comment:
I see we create a `TmpWorkingDir` in `project_config_builder` if the repoDir is "". Would it break anything if we created a working dir even if `len(preWorkflowHooks) == 0` and return an empty working dir so that the ProjectConfigBuilder does not have to worry about the working dir being an empty string? 
 
https://github.com/lyft/atlantis/blob/3b76cc125fef71bfc91681edd792cca788016659/server/neptune/gateway/event/project_config_builder.go#L42-L44

Reply:
I might be misunderstanding what you mean by empty working dir. Do you mean delegate all cloning to the preworkflow `Run()` execution (for both the preworkflow hooks and project config builder)?